include/onnxruntime/core/framework/ortmemoryinfo.h

```diff
@@ -7,7 +7,7 @@ struct OrtMemoryInfo {
   OrtMemoryInfo() = default;  // to allow default construction of Tensor
 
   // use string for name, so we could have customized allocator in execution provider.
-  const char* name = nullptr;
```

Comment:
>const char* name = nullptr; [](http://example.com/codeflow?start=2&length=27)

Why? nullptr clearly shows, the value is not assigned.

Reply:
The ToString member function uses the "name" without checking it is null:

std::string ToString() const {
    std::ostringstream ostr;
    ostr << "OrtMemoryInfo:["
         << "name:" << name
         << " id:" << id
         << " OrtMemType:" << mem_type
         << " OrtAllocatorType:" << alloc_type
         << " " << device.ToString()
         << "]";
    return ostr.str();

Because of this we are seeing memory violation errors in our logs.

Reply:
You may update the ToString() function with something like,
`<< "name:" << name? "" : name`
instead of allocating default value (empty string which is {'\0'}) for it

Reply:
There is not need to update anything. ToString() can only be called on a fully constructed object. There is not a constructor (no default) that would leave the ptr as nullptr. The whole point of -1 and nullptr to be invalid values and signal that the object was not initialized.

Reply:
the struct OrtMemoryInfo also has a default constructor. So the objects can be initialized with default values (name as nullptr). Ideally the ToString should handle nullptr case? Ideally code should not have segfaults?

Reply:
>There is not a constructor (no default) that would leave the ptr as nullptr. 

How about `OrtMemoryInfo() = default; ` ? 

https://github.com/microsoft/onnxruntime/blob/1af4c170ef8dc869ba732043f6f4a38b07338e05/include/onnxruntime/core/framework/ortmemoryinfo.h#L7

Reply:
@yuslepukhin can you please let me know if you still feel the reason is not valid?

Reply:
The default constructed state is not a valid state. I am not seeing any good use of default constucted instance. `nullptr` and related crashes are likely to alert of the fact.

Reply:
We have code which is using the default constructor and then calling ToString API on the instance. The ToString API call is causing memory violation error. So if the default constructor is not a valid state then why give that option? Why not remove the default constructor? This would enforce any user to use the parameterized constructor?

I am just trying to understand that if we are giving the default constructor option then all the member functions should be exception safe. It should not cause any seg faults. If any code outside the member function is directly accessing the nullptr name then we can expect the user to handle it or make sure to initialize it properly.

Reply:
@pash-msft @souptc adding you to the conversation